components/style/properties_and_values.rs

```diff
@@ -0,0 +1,1167 @@
+        Ok(ComputedUrl(url.as_str().to_owned()))
+        // Note: the Gecko binding doesn't currently output resolved URLs.
+    }
+}
+
+impl ToCss for ComputedUrl {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        dest.write_str("url(")?;
+        self.0.to_css(dest)?;
+        dest.write_str(")")
+    }
+}
+
+/// A basic custom property syntax string for a custom property that, used to
+/// build up disjunctions and list terms.
+#[derive(Debug, PartialEq)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum Type {
+    /// Syntax to allow any valid <length> value.
+    Length,
+    /// Syntax to allow any valid <number> value.
+    Number,
+    /// Syntax to allow any valid <percentage> value.
+    Percentage,
+    /// Syntax to allow any valid <length> or <percentage> value, or any valid
+    /// <calc()> expression combining <length> and <percentage> components.
+    LengthPercentage,
+    /// Syntax to allow any valid <color> value.
+    Color,
+    /// Syntax to allow any valid <image> value.
+    Image,
+    /// Syntax to allow any valid <url> value.
+    Url,
+    /// Syntax to allow any valid <integer> value.
+    Integer,
+    /// Syntax to allow any valid <angle> value.
+    Angle,
+    /// Syntax to allow any valid <time> value.
+    Time,
+    // FIXME: We should support <resolution> values as well.
+    /// Syntax to allow any valid <transform-list> value.
+    TransformList,
+    /// Syntax to allow any valid <custom-ident> value.
+    CustomIdent,
+    /// Syntax to allow a specific identifier (matching the <custom-ident>
+    /// production, compared codepoint-wise).
+    SpecificIdent(Ident),
+}
+
+
+impl Type {
+    /// Attempt to parse `input` as this type.
+    pub fn parse<'i, 't>(
+        &self,
+        context: &ParserContext,
+        input: &mut Parser<'i, 't>,
+    ) -> Result<SpecifiedValueItem, StyleTraitsParseError<'i>>
+    {
+        #[cfg(feature = "servo")]
+        fn idents_eq(a: &Atom, b: &str) -> bool {
+            a == b
+        }
+
+        #[cfg(feature = "gecko")]
+        fn idents_eq(a: &Atom, b: &str) -> bool {
+            *a == b.into()
+        }
+
+        macro_rules! parse {
+            ($_self:expr, $context:expr, $input:expr,
+
+             $($typ:ident => $fn:path),*) => {
+                match $_self {
+                    $(
+                        Type::$typ => {
+                            $fn(context, input).map(|x| {
+                                SpecifiedValueItem::$typ(x)
+                            })
+                        }
+                    ), *
+
+                    // We need to actually compare SpecificIdents,
+                    // unfortunately.
+                    Type::SpecificIdent(ref x) => {
+                        $input.expect_ident_cloned().and_then(|y| {
+                            if idents_eq(&**x, &*y) {
```

Comment:
This could use again `CustomIdent::parse($input)`, and then comparing `CustomIdent`s becomes cheap in Gecko too :)

Reply:
Can't seem to find a function with that name: https://doc.servo.org/style/values/struct.CustomIdent.html